common/log.h

```diff
@@ -0,0 +1,617 @@
+    LOG_DISABLE();
+    LOG_SET_TARGET("llama.anotherlog.log");
+    LOG("11 Hello World _2_ to nobody, new target was selected but logs are still disabled!\n")
+    LOG_ENABLE();
+    LOG("12 Hello World this time in a new file ( you should not see _2_ in the log or the output )?\n")
+    LOG_SET_TARGET("llama.yetanotherlog.log");
+    LOG("13 Hello World this time in yet new file?\n")
+    LOG_SET_TARGET(LOG_FILENAME_GENERATOR("llama_autonamed", "log"));
+    LOG("14 Hello World in log with generated filename!\n")
+
+    // exit(0);
+}
+
+inline bool log_param_single_parse(std::string param)
+{
+    std::cerr << "LOGDBG: single param: " << param << std::endl;
+
+    if (std::string("--log-test").compare(param) == 0)
+    {
+        log_test();
+        return true;
+    }
+    else if (std::string("--log-disable").compare(param) == 0)
+    {
+        LOG_DISABLE();
+        return true;
+    }
+    else if (std::string("--log-enable").compare(param) == 0)
+    {
+        LOG_ENABLE();
+        return true;
+    }
+
+    std::cerr << "LOGDBG: single param NO MATCH " << param << std::endl;
+
+    return false;
+}
+
+inline bool log_param_pair_parse(bool check_but_dont_parse, std::string param, std::string next = std::string())
+{
+    std::cerr << "LOGDBG: pair param: " << param << "/" << next << std::endl;
+
+    if (std::string("--log-file").compare(param) == 0)
+    {
+        if (check_but_dont_parse)
+        {
+            return true;
+        }
+        LOG_SET_TARGET(LOG_FILENAME_GENERATOR(next.empty() ? "unnamed" : next, "log"));
+        return true;
+    }
+
+    std::cerr << "LOGDBG: pair param NO MATCH " << param << "/" << next << std::endl;
+
+    return false;
+}
+
+inline void log_print_usage()
+{
+    fprintf(stdout, "log options:\n");
+    /*
+    fprintf(stdout, "  -h, --help            show this help message and exit\n");
+    // spacing
+    fprintf(stdout, "__-param----------------Description\n");*/
+    fprintf(stdout, "  --log-test            Run simple logging test\n");
+    fprintf(stdout, "  --log-disable         Disable trace logs\n");
+    fprintf(stdout, "  --log-enable          Enable trace logs\n");
+    fprintf(stdout, "  --log-file            Specify a log filename (without extension)\n");
+    fprintf(stdout, "                        Log file will be tagged with unique ID and written as \"<name>.<ID>.log\"\n"); /*  */
+}
+
+#ifndef _WIN32
+// TODO:
+//  Windows doesn't seem to like this somehow
+
+#define LOG_DUMP_CMDLINE( argc, argv ) _log_dump_cmdline(argc,argv)
+
+// INTERNAL, DO NOT USE
+inline void _log_dump_cmdline(int argc, char **argv)
+{
+    std::string buf;
+    for (int i = 0; i < argc; ++i)
+    {
+        if (std::string(argv[i]).find(' ') != std::string::npos)
+        {
+            buf.append(" \"").append(argv[i]).append("\"");
+        }
+        else
+        {
+            buf.append(" ").append(argv[i]);
+        }
+    }
+    LOGLN("Cmd:%s", buf.c_str())
```

Comment:
This is how MSVC preprocesses this line:
```cpp
{ if (log_handler() != nullptr) { fprintf(log_handler(), "[%lu]" "[%24s:%5d][%24s] " "Cmd:%s", buf.c_str() "%s" , (std::chrono::duration_cast<std::chrono::duration<std::uint64_t>>(std::chrono::system_clock::now().time_since_epoch())).count() , "llama.cpp\\common\\log.h", 511,  __FUNCTION__  ,"\n"); fflush(log_handler()); } }
```
It doesn't like the `buf.c_str() "%s"` part.

Reply:
Args are unfolded in a completely wrong place, it seems like msvc interprets __VA_ARGS__ differently.

I managed to clean up enough space for windows VM, currently downloading visual studio. I don't want to abuse fragile CI more then necessary, especially after it broke for over half an hour today :)

Reply:
This change seems to fix it:
```diff
-#define LOGLN(...) _LOG(__VA_ARGS__, "\n")
+#define LOGLN(str, ...) _LOG(str, ##__VA_ARGS__, "\n")
```